### PR TITLE
Flutter freezed v4.0.0

### PR DESCRIPTION
**Pull request by Parables**
## Description
The motivation for this version started when @endigo identified some [bugs with the plugin](https://github.com/dotansimha/graphql-code-generator/issues/8330).

As a quick resolution, I suggested [this solution](https://github.com/dotansimha/graphql-code-generator/issues/8330#issuecomment-1292257909) .

Then it hit me... I took a look at the [schema](https://demo.saleor.io/graphql/) @endigo  mentioned and realized how big it was, I tried implementing my initial suggestion on that schema and ended up with a bloated config file spanning close to 500 lines just for some minor configurations.

I was working on a project at that time and was writing pure CSS. It was then that I began to wonder how certain CSS properties have multiple syntaxes. Taking inspiration from the flexibility CSS offers, I began experimenting with RegExp to be used for configuring the plugin.

Initially, I thought I was going too far with this but I stubbornly persisted... +3 months of sleepless nights debugging and tons of refactoring later, comes this PR which hopefully might inspire other plugins too.

Documentation is still in progress and video tutorials will soon follow once this PR gets merged.

Looking forward to your usual collaboration @charlypoly, @Urigo  and all the team at the Guild

Related # (issue)
Replaces #8350 and #14

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

All modules have a corresponding test file that verifies the generated output. 
1. Install the dependencies with `yarrn`
2. run `yarn test`

**Test Environment**:

- OS: Garuda Linux
- `@graphql-codegen/...`: latest
- NodeJS: latest

## Checklist:

- [x] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Parables** [comment]
Sincere gratitude to @auaicn and @endigo and everyone who has helped along the way 🥂

**Comment by Parables** [comment]
@charlypoly @Urigo we have a green flag ... over to you

**Comment by saihaj** [comment]
hey @Parables can you  create a changeset https://github.com/dotansimha/graphql-code-generator-community/pull/47#issuecomment-1383675987

**Comment by Parables** [comment]
Sure, I will that around 10pm GMT

File: .changeset/chatty-teachers-repeat.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        [FieldNamePattern.forFieldNamesOfTypeName([[Droid, id]]), ['default_factory_parameter']],
+      ],
+    }
+```
+
+The 2 configurations above do the same thing, the later being more compact, flexible and readable than the former.
+
+## How to update your existing configuration
+
+First understand the [usage of the Patterns](), then create a new config file(preferably a typescript file: previous version of the code generator used a YAML file).
```
Comment (saihaj) [inline .changeset/chatty-teachers-repeat.md:99]: Did you mean to link some docs here m?

**Comment by Parables** [comment]
@saihaj done as requested

File: .changeset/chatty-teachers-repeat.md (hunk 1 above)
Comment (Parables) [reply .changeset/chatty-teachers-repeat.md:99]: Yes, it will point to the plugin docs on the website

I have another PR for the in the main repo

Comment (Parables) [reply .changeset/chatty-teachers-repeat.md:99]: Fixed

**Comment by Parables** [comment]
@saihaj  I have to correct some typos and grammatical errors in the guide but I have to do that at school as I am already running late. Give me about 1 hour 30 mins to get to school so that we can continue

**Comment by Parables** [comment]
And please how do I fix the prettier error?

**Comment by Parables** [comment]
@saihaj I am back, what else do you need me to do?

**Review by saihaj — APPROVED** [review]

**Comment by Parables** [comment]
> @Parables can you run `yarn prettier` so CI is happy then I can merge and release this.

Done that but there are no changes to be committed here on my local repo

**Comment by saihaj** [comment]
> > @Parables can you run `yarn prettier` so CI is happy then I can merge and release this.
> 
> Done that but there are no changes to be committed here on my local branch.
> 
> However, when I run `yarn` to install the dependencies, other packages were modified.
> 
> I didn't commit them because it was not my intention to modify them.
> 
> Any other solution?

weird I pulled this branch and ran prettier and it was just this plugin https://github.com/dotansimha/graphql-code-generator-community/pull/47/commits/b6c4b0159c53530251dc50040eeac282f6e128e9

**Comment by Parables** [comment]
Thanks very much @saihaj. I couldn't have figured out this issue on my own.

The guide for this plugin on the website to has been updated.

Everything is ready now for release.

**Comment by saihaj** [comment]
hmm so something went wrong with CI publish flow. So will investigate it @Parables

**Comment by Parables** [comment]
Thanks very much for your dedication @saihaj

**Comment by saihaj** [comment]
released @Parables https://github.com/dotansimha/graphql-code-generator-community/releases/tag/release-1674417409150

**Comment by Parables** [comment]
Thank you all so much for the support.

This release wouldn't be possible without you:

@saihaj 
@charlypoly 
@Urigo 
@endigo
@n1ru4l 
@auaicn
@gabsn 
 @everyone

Cheers 🥂

**Comment by asanwura** [comment]
> hmm so something went wrong with CI publish flow. So will investigate it @Parables

I'm curious to learn why it failed. a little insight could help me learn from my mistakes

**Comment by Parables** [comment]
Thank you all so much for the support.

This release wouldn't be possible without you:

@saihaj 
@charlypoly 
@Urigo 
@endigo
@n1ru4l 
@auaicn
@gabsn 
 @everyone

Cheers 🥂

**Pull request merged**